Agora vamos realizar o desafio do módulo 2, que sinceramente, quero ver como eu vou fazer esse trem complicado, <strong>JESUS</strong>


In [2]:
#Vamos começar importando o que eu lembro, pelo menos 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from google.colab import files

In [35]:
df = pd.read_csv("https://pycourse.s3.amazonaws.com/banknote_authentication.txt", header=None, names=['variance', 'sknewness', 'curtosis', 'entropy', 'class'])
#Lendo essa poha de df

In [36]:
#Vamos dar uma analisada nesse dataframe
print(df.head())
print('')
df.info()

   variance  sknewness  curtosis  entropy  class
0   3.62160     8.6661   -2.8073 -0.44699      0
1   4.54590     8.1674   -2.4586 -1.46210      0
2   3.86600    -2.6383    1.9242  0.10645      0
3   3.45660     9.5228   -4.0112 -3.59440      0
4   0.32924    -4.4552    4.5718 -0.98880      0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   variance   1372 non-null   float64
 1   sknewness  1372 non-null   float64
 2   curtosis   1372 non-null   float64
 3   entropy    1372 non-null   float64
 4   class      1372 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 53.7 KB


In [37]:
#Agora que demos uma analisada, vamos separar nossos dados em treino e teste que eu sinceramente, acho que deveria transformar essa class em categorico
df['class'].value_counts()

0    762
1    610
Name: class, dtype: int64

In [38]:
#Vou cirar um dataframe alterado para poder mexer
df_alterado = df.copy()
print(df_alterado)
df_alterado['class'] = df_alterado['class'].astype('category')
print(df_alterado.info())
df_alterado['class']

      variance  sknewness  curtosis  entropy  class
0      3.62160    8.66610   -2.8073 -0.44699      0
1      4.54590    8.16740   -2.4586 -1.46210      0
2      3.86600   -2.63830    1.9242  0.10645      0
3      3.45660    9.52280   -4.0112 -3.59440      0
4      0.32924   -4.45520    4.5718 -0.98880      0
...        ...        ...       ...      ...    ...
1367   0.40614    1.34920   -1.4501 -0.55949      1
1368  -1.38870   -4.87730    6.4774  0.34179      1
1369  -3.75030  -13.45860   17.5932 -2.77710      1
1370  -3.56370   -8.38270   12.3930 -1.28230      1
1371  -2.54190   -0.65804    2.6842  1.19520      1

[1372 rows x 5 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   variance   1372 non-null   float64 
 1   sknewness  1372 non-null   float64 
 2   curtosis   1372 non-null   float64 
 3   entropy    1372 non-null   float64 

0       0
1       0
2       0
3       0
4       0
       ..
1367    1
1368    1
1369    1
1370    1
1371    1
Name: class, Length: 1372, dtype: category
Categories (2, int64): [0, 1]

In [39]:
#Agora vou importar do sklearn o train test split
from sklearn.model_selection import train_test_split

In [40]:
#Agora que eu importei, vou dividir o df alterado em treino e teste
xtrain, xtest, ytrain, ytest = train_test_split(df_alterado.iloc[:, :-1], df_alterado['class'], test_size = 0.3, random_state = 1)

In [41]:
#Vamos dar uma verificada se eu consegui separar direito
print(xtrain)
print(ytrain)

      variance  sknewness  curtosis   entropy
694    0.74054    0.36625  2.199200  0.484030
880   -2.35180   -4.83590  6.647900 -0.060358
1251  -0.77688    0.13036 -0.031137 -0.353890
812   -2.06590    1.05120 -0.462980 -1.097400
1212  -2.45600   -0.24418  1.404100 -0.458630
...        ...        ...       ...       ...
715    3.49160    8.57090 -3.032600 -0.591820
905    0.74521    3.63570 -4.404400 -4.141400
1096  -4.36670    6.06920  0.572080 -5.466800
235    2.04660    2.03000  2.176100 -0.083634
1061  -2.31470    3.66680 -0.696900 -1.247400

[960 rows x 4 columns]
694     0
880     1
1251    1
812     1
1212    1
       ..
715     0
905     1
1096    1
235     0
1061    1
Name: class, Length: 960, dtype: category
Categories (2, int64): [0, 1]


In [ ]:
#Beleza, até o momento fiz tudo certo, Agora, antes de aplicarmos nos modelos, como temos nosso ambiente pronto, vamos responder
#as perguntas do desafio

1. Qual o tamanho desse dataset (número de linhas, número de colunas)?<br>R:(1372,5)

In [44]:
df.info

<bound method DataFrame.info of       variance  sknewness  curtosis  entropy  class
0      3.62160    8.66610   -2.8073 -0.44699      0
1      4.54590    8.16740   -2.4586 -1.46210      0
2      3.86600   -2.63830    1.9242  0.10645      0
3      3.45660    9.52280   -4.0112 -3.59440      0
4      0.32924   -4.45520    4.5718 -0.98880      0
...        ...        ...       ...      ...    ...
1367   0.40614    1.34920   -1.4501 -0.55949      1
1368  -1.38870   -4.87730    6.4774  0.34179      1
1369  -3.75030  -13.45860   17.5932 -2.77710      1
1370  -3.56370   -8.38270   12.3930 -1.28230      1
1371  -2.54190   -0.65804    2.6842  1.19520      1

[1372 rows x 5 columns]>

2. Qual variável possui o maior range (diferença entre valor máximo e mínimo)?
<br> R:Sknewness

In [49]:
print(df.describe())
ranges = df.max() - df.min()
print(ranges)

          variance    sknewness     curtosis      entropy        class
count  1372.000000  1372.000000  1372.000000  1372.000000  1372.000000
mean      0.433735     1.922353     1.397627    -1.191657     0.444606
std       2.842763     5.869047     4.310030     2.101013     0.497103
min      -7.042100   -13.773100    -5.286100    -8.548200     0.000000
25%      -1.773000    -1.708200    -1.574975    -2.413450     0.000000
50%       0.496180     2.319650     0.616630    -0.586650     0.000000
75%       2.821475     6.814625     3.179250     0.394810     1.000000
max       6.824800    12.951600    17.927400     2.449500     1.000000
variance     13.8669
sknewness    26.7247
curtosis     23.2135
entropy      10.9977
class         1.0000
dtype: float64


3. Qual a média da coluna skewness?
<br> R: 1.92235

4. Qual a média da coluna entropy?
<br> R: -1.191657

5. Qual a mediana da coluna variance?
<br> R:0.49618

In [50]:
#Para o calculo da mediana :
df['variance'].median()

0.49618

6. Qual a desvio padrão da coluna curtosis?
<br> R: 4,310030

7. Qual a porcentagem de exemplos do dataset que são cédulas falsas (class=1)?
<br> R:44.46%

In [53]:
df[df['class'] == 1].count()

variance     610
sknewness    610
curtosis     610
entropy      610
class        610
dtype: int64

In [56]:
print(610/len(df)*100)

44.460641399416915


8. Qual o valor da correlação de Pearson entre as variáveis skewness e curtosis?
<br> R: -0.786895

In [57]:
df[['sknewness', 'curtosis']].corr()

sknewness  curtosis
sknewness   1.000000 -0.786895
curtosis   -0.786895  1.000000

9. Qual a acurácia do KNN no conjunto de teste?
<br> R: 0,9975728

In [58]:
#Agora vamos importar as bibliotecas e os modelos de aprendizagem
from sklearn.neighbors import KNeighborsClassifier
clf_KNN = KNeighborsClassifier(n_neighbors = 5)

In [59]:
from sklearn.tree import DecisionTreeClassifier
clf_arvore = DecisionTreeClassifier(random_state = 1)

In [60]:
from sklearn.ensemble import RandomForestClassifier
clf_floresta = RandomForestClassifier(max_depth = 8, random_state = 1)

In [61]:
from sklearn.svm import SVC
clf_svm = SVC(gamma ='auto',kernel='rbf',random_state=1)

In [64]:
from sklearn.neural_network import MLPClassifier
clf_mlp = MLPClassifier(hidden_layer_sizes=(2,), solver='lbfgs', random_state=1)

In [65]:
#Agora que importei tudo, temos de treinar os modelos
clf_KNN.fit(xtrain, ytrain)
ypredi_knn = clf_KNN.predict(xtest)

In [66]:
#Agora que treinei o modelo, vamos medir a acurácia
from sklearn.metrics import accuracy_score
accuracy_score(ytest, ypredi_knn)

0.9975728155339806

10. Qual a acurácia da Árvore de Decisão no conjunto de teste?
<br> R: 0.98786407767

In [67]:
#Vamos treinar a árvore de decisão e ver quais o seus resultados
clf_arvore.fit(xtrain,ytrain)
ypredic_arvore = clf_arvore.predict(xtest)

In [68]:
#Agora vamos medir a acurácia do modelo
accuracy_score(ytest,ypredic_arvore)

0.9878640776699029

11. Qual a acurácia do Random Forest no conjunto de teste?
<br> R: 0.995145631

In [69]:
#Vamos fazer como já vinhamos fazendo, treinar o modelo e analisar a acurácia
clf_floresta.fit(xtrain,ytrain)
ypredic_floresta = clf_floresta.predict(xtest)

In [70]:
#Medindo a acurácia
accuracy_score(ytest,ypredic_floresta)

0.9951456310679612

12. Analisando o valor da importância relativa das features do Random Forest (atributo feature_importances_), qual feature melhor contribuiu para a predição de class?
<br> R: 

In [75]:
#Vamos aplicar o atributo e ver qual a saída
print(clf_floresta.feature_importances_)
xtrain.columns

[0.56066718 0.24058589 0.14022782 0.05851911]


Index(['variance', 'sknewness', 'curtosis', 'entropy'], dtype='object')

13. Qual o valor da importância relativa da feature skewness?
<br> R: 0.24058589

14. Qual a acurácia do SVM no conjunto de teste?
<br> R: 1.0 (WHAT???)

In [76]:
#Vamos treinar o modelo assim como fizemos com os outros
clf_svm.fit(xtrain,ytrain)
ypredic_svm = clf_svm.predict(xtest)

In [77]:
#Medindo a acurácia
accuracy_score(ytest, ypredic_svm)

1.0

15. Qual a acurácia da rede MLP no conjunto de teste?
<br> R:1.0

In [78]:
#Vamos a rede MLP assim como os outros e medir sua acurácia
clf_mlp.fit(xtrain, ytrain)
ypredic_mlp = clf_mlp.predict(xtest)

In [79]:
#Medindo a acurácia da MLP
accuracy_score(ytest, ypredic_mlp)

1.0